pkg/scheduler/queue/tenant_queues.go

```diff
@@ -23,32 +24,66 @@ type tenantRequest struct {
 // queueBroker encapsulates access to tenant queues for pending requests
 // and maintains consistency with the tenant-querier assignments
 type queueBroker struct {
-	tenantQueuesTree *TreeQueue
+	tree *TreeQueue
 
-	tenantQuerierAssignments tenantQuerierAssignments
+	tenantQuerierAssignments *tenantQuerierAssignments
 
 	maxTenantQueueSize               int
 	additionalQueueDimensionsEnabled bool
+	prioritizeQueryComponents        bool
 }
 
-func newQueueBroker(maxTenantQueueSize int, additionalQueueDimensionsEnabled bool, forgetDelay time.Duration) *queueBroker {
-	return &queueBroker{
-		tenantQueuesTree: NewTreeQueue("root"),
-		tenantQuerierAssignments: tenantQuerierAssignments{
-			queriersByID:       map[QuerierID]*querierConn{},
-			querierIDsSorted:   nil,
-			querierForgetDelay: forgetDelay,
-			tenantIDOrder:      nil,
-			tenantsByID:        map[TenantID]*queueTenant{},
-			tenantQuerierIDs:   map[TenantID]map[QuerierID]struct{}{},
-		},
+func newQueueBroker(
+	maxTenantQueueSize int,
+	additionalQueueDimensionsEnabled bool,
+	prioritizeQueryComponents bool,
+	forgetDelay time.Duration,
+) *queueBroker {
+	currentQuerier := QuerierID("")
+	tqas := &tenantQuerierAssignments{
+		queriersByID:       map[QuerierID]*querierConn{},
+		querierIDsSorted:   nil,
+		querierForgetDelay: forgetDelay,
+		tenantIDOrder:      nil,
+		tenantsByID:        map[TenantID]*queueTenant{},
+		tenantQuerierIDs:   map[TenantID]map[QuerierID]struct{}{},
+		currentQuerier:     &currentQuerier,
+		tenantOrderIndex:   localQueueIndex - 1,
+	}
+
+	// If prioritizeQueryComponents is true, the tree will be created
```

Comment:
while this isn't a functional change, I *think* the pattern we follow now is to use the slow rollouts to introduce something with the feature flag on and let the rollout pattern slowly expose issues (if any), then the flag can be used to turn something off in a hurry if needed.

this would mean you could punt the flag (and tests for when the flag is on) until next PR, as I don't think we should merge any config that doesn't have tests for it being on, even if _we_ plan to not turn it on yet

Reply:
I've removed the flag from the config for now, but left this argument (hardcoded) to lay the groundwork for when we re-introduce the config flag.